src/Chainweb/Chainweb/MinerResources.hs

```diff
@@ -85,24 +95,96 @@ withMiningCoordination
+                , _coord503s = c503
+                , _coord403s = c403
+                , _coordConf = conf
+                , _coordUpdateStreamCount = l
+                , _coordPrimedWork = m })
   where
+    cids :: [ChainId]
+    cids = HS.toList . chainIds $ _chainwebVersion cdb
+
+    -- | THREAD: Keep a live-updated cache of Payloads for specific miners, such
+    -- that when they request new work, the block can be instantly constructed
+    -- without interacting with the Pact Queue.
+    --
+    primeWork :: [Miner] -> TVar PrimedWork -> Cut -> ChainId -> IO ()
+    primeWork miners tpw c cid = runForever (logFunction logger) "primeWork" $ go c
+      where
+        go :: Cut -> IO a
+        go cut = do
+            -- Even if the `cut` passed to the function is old, this call here will
+            -- immediately detect the newest `BlockHeader` on the given chain.
+            new <- awaitNewCutByChainId cdb cid cut
+            -- Temporarily block this chain from being considered for queries --
+            atomically $ modifyTVar' tpw (silenceChain cid)
+            -- Generate new payloads, one for each Miner we're managing --
+            let !newParent = ParentHeader . fromJuste . HM.lookup cid $ _cutMap new
+            payloads <- traverse (\m -> T2 m <$> getPayload newParent m) miners
+            -- Update the cache in a single step --
+            atomically $ modifyTVar' tpw (\pw -> foldl' (updateCache cid) pw payloads)
+            go new
+
+    -- | Declare that a particular Chain is temporarily unavailable for new work
+    -- requests while a new payload is being formed.
+    --
+    silenceChain :: ChainId -> PrimedWork -> PrimedWork
+    silenceChain cid (PrimedWork pw) = PrimedWork (pw & traverse . at cid ?~ Nothing)
+
+    updateCache
+        :: ChainId
+        -> PrimedWork
+        -> T2 Miner (T2 PayloadWithOutputs BlockCreationTime)
+        -> PrimedWork
+    updateCache cid (PrimedWork pw) (T2 (Miner mid _) payload) =
+        PrimedWork (pw & at mid . traverse . at cid ?~ Just payload)
+
+    initialPayloads :: Cut -> [Miner] -> IO PrimedWork
+    initialPayloads cut ms =
+        PrimedWork . HM.fromList <$> traverse (\m -> (view minerId m,) <$> fromCut m pairs) ms
+      where
+        pairs :: [T2 ChainId ParentHeader]
+        pairs = map (\cid -> T2 cid (ParentHeader (cut ^?! ixg cid))) cids
+
+    -- | Based on a `Cut`, form payloads for each chain for a given `Miner`.
+    --
+    fromCut
+      :: Miner
+      -> [T2 ChainId ParentHeader]
+      -> IO (HM.HashMap ChainId (Maybe (T2 PayloadWithOutputs BlockCreationTime)))
+    fromCut m cut =
+        HM.fromList <$> traverse (\(T2 cid bh) -> (cid,) . Just <$> getPayload bh m) cut
+
+    getPayload :: ParentHeader -> Miner -> IO (T2 PayloadWithOutputs BlockCreationTime)
+    getPayload (ParentHeader parent) m = do
+        creationTime <- BlockCreationTime <$> getCurrentTimeIntegral
+        payload <- _pactNewBlock pact m parent creationTime
```

Comment:
@larskuhtz This isn't `trace`d like the original call. Should it be?

Reply:
Yes, please trace all calls of `_pactNewBlock`. These is important information for measuring overall performance of the system and catching performance regressions.

You may provide the full qualified path of the function. That would allow to distinguish the different code paths in the telemetry data, if needed.